File: frontend/src/services/PurchaseService.ts
import { UserProfile } from '../types';

export class PurchaseService {
    /**
     * Validates if a purchase is possible.
     */
    static canAfford(userProfile: UserProfile, cost: number): boolean {
        return userProfile.coins >= cost;
    }

    /**
     * Processes the transaction logic (calculating new balance).
     * Does not update state directly, returns new balance.
     */
    static processTransaction(currentCoins: number, cost: number): number {
        if (currentCoins < cost) {
            throw new Error("Insufficient funds");
        }
        return currentCoins - cost;
    }
    
    // Receipt generation mentioned in prompt
    static generateReceipt(itemId: string, cost: number, remainingCoins: number): string {
        return `Purchased ${itemId} for ${cost}. Remaining balance: ${remainingCoins}`;
    }
}

File: mobile/lib/services/purchase_service.dart
class PurchaseService {
  /// Check if user can afford the item.
  static bool canAfford(int coins, int cost) {
    return coins >= cost;
  }

  /// Process transaction and return new coin balance.
  /// Throws Exception if insufficient funds.
  static int processTransaction(int currentCoins, int cost) {
    if (currentCoins < cost) {
      throw Exception('Insufficient funds');
    }
    return currentCoins - cost;
  }

  /// Generate receipt string.
  static String generateReceipt(String itemId, int cost, int remainingCoins) {
    return 'Purchased $itemId for $cost. Remaining balance: $remainingCoins';
  }
}
